RFC: Extract all parts of QR factorization from SPQR

andreasnoack opened the pull request:
I think this is mostly done. The documentation needs to be adjusted a bit.

Instead of storing a void pointer to a SPQR object, this PR extracts the Householder reflectors and loadings, the R factor and the permutations produced by SPQR and store them in the Julia type `SparseQR`. This makes the sparse QR more similar to the other QRs. The SPQR C API doesn't allow you to return these components to the library (as far as I can tell) so a consequence of this change is that Q multiplication and least squares solve must be implemented in Julia. For now, I've implemented the basic solution in the case of rank deficiency. We can add the minimum norm solution later.

cc: @dpo

Update:
The PR includes some general improvements to some functions acting on column views of a sparse matrix. Some the existing methods for sparse matrices require very limited adjustments to become efficient for column views. These changes are pretty selfcontained so I might factor them out of this PR.

andreasnoack committed:
Wrap full SuiteSparseQR function

andreasnoack committed:
Introduce new QRSparse type and write Q functions in Julia

andreasnoack committed:
Remove unused code and adjust tests

tkelman commented on test/sparse/spqr.jl:
```diff
@@ -27,34 +60,18 @@ for eltyA in (Float64, Complex{Float64})
     end
 end
 
-# Issue 14134
```
are you intentionally deleting this test?

andreasnoack replied:
Yes. The test is related to the handling of pointers to SPQR objects but with this PR, we don't have such pointers anymore. All the data is now stored in Julia structs.

tkelman commented on base/sparse/spqr.jl:
```diff
@@ -174,11 +265,44 @@ function (\)(F::Factorization{Float64}, B::VecOrMat{Complex{Float64}})
+        end
+    end
+
+    # Make a view into x correstonding to the size of b
```
corresponding

tkelman commented on base/sparse/spqr.jl:
```diff
@@ -25,122 +25,213 @@ const ORDERING_BESTAMD = Int32(9) # try COLAMD and AMD; pick best#
-    factors::Ptr{Tv}
-end
+function _qr!(ordering, tol, econ, getCTX, A::Ptr{CHOLMOD.C_Sparse{Tv}},
+    Bsparse::Ptr{CHOLMOD.C_Sparse{Tv}}        = Ptr{CHOLMOD.C_Sparse{Tv}}(C_NULL),
```
The indentation here makes it look like these lines are part of the body of the function. An extra indent level for line continuations of block openers would be clearer here IMO.

tkelman commented on base/sparse/sparsevector.jl:
```diff
@@ -25,15 +25,33 @@ end
+function nonzeros(x::SubArray{T,1,<:AbstractSparseArray,Tuple{Base.Slice{Base.OneTo{Int64}},Int64},false} where T)
+    rowidx, colidx = parentindexes(x)
+    A = parent(x)
+    @inbounds y = view(A.nzval, A.colptr[colidx]:A.colptr[colidx + 1] - 1)
```
`parent` isn't always guaranteed to have a `.nzval` or `.colptr` field if it's only required to be an `AbstractSparseArray`

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -27,6 +27,21 @@ end
 
 size(S::SparseMatrixCSC) = (S.m, S.n)
 
+const SparseMatrixCSCUnion{Tv,Ti} =
```
could use some comments exactly what the SubArray part of this is covering

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -27,6 +27,21 @@ end
 
+const SparseMatrixCSCUnion{Tv,Ti} =
+    Union{SparseMatrixCSC{Tv,Ti},
+        SubArray{Tv,2,SparseMatrixCSC{Tv,Ti},Tuple{Base.Slice{Base.OneTo{Int64}},I}}} where {I<:AbstractUnitRange}
```
Slice and OneTo aren't imported in this module?

andreasnoack replied:
I don't think they are

tkelman commented on base/sparse/spqr.jl:
```diff
@@ -25,122 +25,213 @@ const ORDERING_BESTAMD = Int32(9) # try COLAMD and AMD; pick best#
+    rpivinv::Vector{Ti}
+end
+
+Base.size(F::QRSparse) = (size(F.factors, 1), size(F.R, 2))
```
is size not already imported?

tkelman commented on base/sparse/spqr.jl:
```diff
@@ -25,122 +25,213 @@ const ORDERING_BESTAMD = Int32(9) # try COLAMD and AMD; pick best#
-        throw(DimensionMismatch("left hand side and right hand side must have same number of rows"))
+Base.size(Q::QRSparseQ) = (size(Q.factors, 1), size(Q.factors, 1))
+
+function Base.LinAlg.qrfact(A::SparseMatrixCSC{Tv}, ::Type{Val{true}}; tol = size(A, 1)*eps()) where {Tv <: CHOLMOD.VTypes}
```
don't need the `Base.LinAlg.` since `qrfact` is already imported

andreasnoack committed:
Update docstrings and add some comments

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -27,6 +27,10 @@ end
 size(S::SparseMatrixCSC) = (S.m, S.n)
 
+# Define a union of SparseMatrixCSC and view containing all rows of a SparseMatrixCSC and a contiguous
+# selection of columns. Many methods can defined efficiently for this union by extracting the fields
```
can be defined

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -27,6 +27,10 @@ end
+# Define a union of SparseMatrixCSC and view containing all rows of a SparseMatrixCSC and a contiguous
+# selection of columns. Many methods can defined efficiently for this union by extracting the fields
+# via the get function: getcolptr, getrowval, and getnzval. The key insight is that getcolptr on a view
+# returns a offset view of the colptr of the underlying SparseMatriCSC
```
SparseMatrixCSC

tkelman commented on base/sparse/spqr.jl:
```diff
@@ -233,12 +233,29 @@ end
-exported.
+Compute the `QR` factorization of a sparse matrix `A`. Fill-reducing row and columns permutations
+are used such that `F[:R] = F[:Q]'*A[F[:prow],F[:pcol]]`. The main application of this type is to
+solve least squares or underdertermined problems with [`\\`](@ref). The function calls the C library SPQR.
```
underdetermined

tkelman commented on base/sparse/spqr.jl:
````diff
@@ -233,12 +233,29 @@ end
+are used such that `F[:R] = F[:Q]'*A[F[:prow],F[:pcol]]`. The main application of this type is to
+solve least squares or underdertermined problems with [`\\`](@ref). The function calls the C library SPQR.
+
+```jldoctest
````
```
# Examples
```jldoctest
```

tkelman commented on base/sparse/spqr.jl:
````diff
@@ -303,6 +320,26 @@ function _ldiv_basic(F::QRSparse, B::StridedVecOrMat)
+``Ax=b`` when `F` is the sparse QR factorization of ``A``. A basic solution is returned
+when the problem is underdetermined.
+
+```jldoctest
````
```
# Examples
```jldoctest
```

Sacha0 commented on base/sparse/sparsematrix.jl:
```diff
@@ -27,6 +27,25 @@ end
+# Define a union of SparseMatrixCSC and view containing all rows of a SparseMatrixCSC and a contiguous
+# selection of columns. Many methods can be defined efficiently for this union by extracting the fields
+# via the get function: getcolptr, getrowval, and getnzval. The key insight is that getcolptr on a view
+# returns a offset view of the colptr of the underlying SparseMatrixCSC
```
"returns a offset view" -> "returns an offset view"?

Sacha0 commented on base/sparse/sparsematrix.jl:
```diff
@@ -27,6 +27,25 @@ end
+    S.parent.rowval
+getnzval(S::SparseMatrixCSC) = S.nzval
+getnzval(S::SubArray{T,2,<:SparseMatrixCSC,Tuple{Base.Slice{Base.OneTo{Int64}},I}} where {T,I<:AbstractUnitRange}) =
+    S.parent.nzval
```
Lovely! :)

Perhaps worth giving this form of view a name, to avoid repeating the rather complex definition?

Sacha0 commented on base/sparse/sparsevector.jl:
```diff
@@ -25,15 +25,33 @@ end
 
+const AbstractSparseVectorUnion{T<:Number} =
+    Union{AbstractSparseVector{T},
+          SubArray{T,1,<:AbstractSparseArray,Tuple{Base.Slice{Base.OneTo{Int64}},Int64},false}}
```
Perhaps `<:AbstractSparseMatrix` rather than `<:AbstractSparseArray`? The `Tuple` defining the slice appears specific to two dimensional arrays. (Perhaps even just `<:SparseMatrixCSC`? The `AbstractSparse[Array|Matrix|Vector]` interfaces are not well defined, and the methods below appear to be written against the `SparseMatrixCSC` interface.)

Sacha0 commented on base/sparse/sparsevector.jl:
```diff
@@ -25,15 +25,33 @@ end
+function nonzeros(x::SubArray{T,1,<:SparseMatrixCSC,Tuple{Base.Slice{Base.OneTo{Int64}},Int64},false} where T)
+    rowidx, colidx = parentindexes(x)
+    A = parent(x)
+    @inbounds y = view(A.nzval, A.colptr[colidx]:A.colptr[colidx + 1] - 1)
```
Could the `A.colptr[...` simplify to `nzrange(A, colidx)`? Likewise below.

Sacha0 commented on base/sparse/spqr.jl:
```diff
@@ -25,129 +25,238 @@ const ORDERING_BESTAMD = Int32(9) # try COLAMD and AMD; pick best#
-    qrfact(A) -> SPQR.Factorization
+    qrfact(A) -> QRSparse
+
+Compute the `QR` factorization of a sparse matrix `A`. Fill-reducing row and columns permutations
```
"row and columns permutations" -> "row and column permutations"?

andreasnoack committed:
Define some aliases, fix nonzeros for views, and some minor fixes.

andreasnoack committed:
Fix ambiguities and finish docstring for getindex

tkelman commented:
segfault on 32 bit looks like a real problem. worth splitting out the view aliases to a different pr in the meantime?

andreasnoack commented:
Yes. It is a real problem. I can reproduce locally if I run the factorization in a loop. Still investigating. Will factorize out the sparse alias stuff if I can't fix this soon.

andreasnoack committed:
Fix segfault by avoiding that Sparse object is getting freed while calling
SPQR. This is accomplished by passing Sparse and define unsafe_convert instead
of just passing the pointer.

Also fix memory leak.

andreasnoack commented:
Now without segfault and memory leaks. I think this one is ready.

andreasnoack committed:
New ntuple use

tkelman commented:
32 bit test failures are in spqr

tkelman commented on base/sparse/sparsevector.jl:
```diff
@@ -30,15 +30,34 @@ end
 
+# Define an alias for a view of a whole column of a SparseMatrixCSC. Many methods can be written for the
+# union of such a view and a SparseVector so we define an alias for such a union as well
+const SparseVectorView{T}  = SubArray{T,1,<:SparseMatrixCSC,Tuple{Base.Slice{Base.OneTo{Int64}},Int64},false}
```
maybe call this SparseColumnView? since it's not a view **of** a SparseVector, but rather a view of a single column of a sparse matrix **as** as a SparseVector

tkelman commented on base/sparse/sparsevector.jl:
```diff
@@ -1518,6 +1536,19 @@ end
 ### BLAS-2 / dense A * sparse x -> dense y
 
+# lowrankupdate (BLAS.ger! like)
+function LinAlg.lowrankupdate!(A::StridedMatrix, x::StridedVector, y::SparseVectorUnion, α::Number = 1)
```
not consistent with the dense method if any elements of `x` are non-finite

andreasnoack replied:
Most our sparse linear algebra (including CHOLMOD) has this behavior right now. E.g.
```julia
julia> Base.SparseArrays.CHOLMOD.Sparse(speye(5))*Base.SparseArrays.CHOLMOD.Dense([Inf; ones(4)])
5×1 Base.SparseArrays.CHOLMOD.Dense{Float64}:
 Inf
   1.0
   1.0
   1.0
   1.0
```
I think we should just not follow IEEE here (which also BLAS deviates from) but I know that you don't agree. In any case, changing the behavior of our sparse linear algebra is a separate issue.

tkelman replied:
we can put in a check and call the slow dense fallback, can't we?

tkelman replied:
broadcast, and any generic dense implementations, would handle the corner cases here. We don't need to repeat other libraries' bugs when it's a fairly straightforward (and not that expensive) check to do the slow but right thing in rare cases, instead of pretending the cases don't exist and ignoring them.

andreasnoack replied:
We can continue this one in https://github.com/JuliaLang/julia/issues/22733

tkelman commented on base/sparse/spqr.jl:
```diff
@@ -21,133 +21,293 @@ const ORDERING_BESTAMD = Int32(9) # try COLAMD and AMD; pick best#
 # tried.  If there is a high fill-in with AMD then try METIS(A'A) and take
```
We've never been building suitesparse with metis support - older versions of metis had a strange license.

andreasnoack replied:
I don't really remember why I included this comment. I don't think it is useful so I'll just delete it.

andreasnoack replied:
Well, actually it describes the default behavior so I guess it is okay to keep it. In theory, users could change the build options for SuiteSparse and in any case the comment is still correct, right?

tkelman commented on base/sparse/spqr.jl:
```diff
@@ -21,133 +21,293 @@ const ORDERING_BESTAMD = Int32(9) # try COLAMD and AMD; pick best#
-        end
-        new(m, n, p)
-    end
+# Such that A[invperm(p), q] = (I - H[:,1]*τ[1]*H[:,1]')*...*(I - H[:,k]*τ[k]*H[:,k]')*R
```
is the start of this comment missing?

tkelman commented on base/sparse/spqr.jl:
```diff
@@ -21,133 +21,293 @@ const ORDERING_BESTAMD = Int32(9) # try COLAMD and AMD; pick best#
+    elseif i > 2
+        return 1
+    else
+        throw(ArgumentError("second argument must be 1 or 2"))
```
error seems inconsistent with allowed inputs, any strictly positive integer is accepted

tkelman commented on base/sparse/spqr.jl:
```diff
@@ -21,133 +21,293 @@ const ORDERING_BESTAMD = Int32(9) # try COLAMD and AMD; pick best#
-    d
+Base.size(Q::QRSparseQ) = (size(Q.factors, 1), size(Q.factors, 1))
+
+# From SPQR manula p. 6
```
manual

tkelman commented on base/sparse/spqr.jl:
```diff
@@ -21,133 +21,293 @@ const ORDERING_BESTAMD = Int32(9) # try COLAMD and AMD; pick best#
+  [4, 2]  =  1.0
+
+julia> qrfact(A)
+Base.SparseArrays.SPQR.QRSparse{Float64,Int64}(
```
should there be a show method that prints the factors nicely?

tkelman commented on base/sparse/spqr.jl:
```diff
@@ -21,133 +21,293 @@ const ORDERING_BESTAMD = Int32(9) # try COLAMD and AMD; pick best#
-        throw(DimensionMismatch("Q matrix size $mQR and dense matrix has $mX rows"))
-    elseif (method == XQT || method == XQ) && mQR != nX
-        throw(DimensionMismatch("Q matrix size $mQR and dense matrix has $nX columns"))
+function Base.Ac_mul_B!(Q::QRSparseQ, A::StridedVecOrMat)
```
how do these compare perf-wise to ccall'ing the mult methods from spqr?

andreasnoack replied:
I did a few comparisons and the Julia implementations are faster. Probably because they avoid the conversion to `CHOLMOD.Sparse`.

tkelman commented on base/sparse/spqr.jl:
```diff
@@ -21,133 +21,293 @@ const ORDERING_BESTAMD = Int32(9) # try COLAMD and AMD; pick best#
+    getindex(F::QRSparse, d::Symbol)
+
+Extract factors of a QRSparse factorization. Possible values of `d` are
+- :Q : `QRSparseQ` matrix of the ``Q`` factor in Householder form
```
should probably code highlight the `:Q` etc symbols

tkelman commented on base/sparse/spqr.jl:
```diff
@@ -174,11 +334,65 @@ function (\)(F::Factorization{Float64}, B::VecOrMat{Complex{Float64}})
+    end
+
+    # The rank of F equal to the number of rows in R
+    rnk = size(F.R,1)
```
(minor) `size(F.R, 1)`

tkelman commented on base/sparse/spqr.jl:
```diff
@@ -174,11 +334,65 @@ function (\)(F::Factorization{Float64}, B::VecOrMat{Complex{Float64}})
+        end
+    end
+
+    # Make a view into x corresponding to the size of B
```
capital X here and in "extract solution from" below, right?

tkelman commented on test/sparse/spqr.jl:
```diff
@@ -27,34 +60,18 @@ for eltyA in (Float64, Complex{Float64})
         @test A[1:9,:]*(A[1:9,:]\ones(eltyB, 9)) ≈ ones(9) # Underdetermined system
-
-        if eltyA == eltyB # promotions not defined for unexported methods
-            @test qrfact(sparse(eye(eltyA, 5)))\ones(eltyA, 5) == ones(5)
```
this one should still work, right?

andreasnoack replied:
True. I've added it.

andreasnoack committed:
Add show method for QRSparse

Minor adjustments following review recommendations

tkelman commented on base/sparse/spqr.jl:
```diff
@@ -309,6 +324,18 @@ function Base.getindex(F::QRSparse, d::Symbol)
 end
 
+function Base.show(io::IO, mime::MIME{Symbol("text/plain")}, F::QRSparse)
+    println(io, summary(F))
```
not implemented?

andreasnoack replied:
I don't understand the question.

tkelman replied:
there is no summary method for this type, it doesn't output anything

andreasnoack replied:
It is using [hits fallback](https://github.com/JuliaLang/julia/blob/master/base/show.jl#L1606) to write a string with the type information

tkelman commented on base/sparse/spqr.jl:
```diff
@@ -309,6 +324,18 @@ function Base.getindex(F::QRSparse, d::Symbol)
+    show(io, mime, F[:R])
+    println(io, "\nRow permutation:")
+    show(io, mime, F[:prow])
+    println(io, "\nColumns permutation:")
```
shouldn't be plural if "Row permutation" isn't
